common/src/primitives/merkle/tree/tree_size.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    fmt::{Display, Formatter},
+    num::NonZeroUsize,
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TreeSize(usize);
```

Comment:
I'm staring at the implementations in this file and thinking how this will help. I don't get it. The only handling is in constructors... isn't this the point? Moreover, if we use a low vile like the log2, then we have to check (also) that it's not too high.

Reply:
The idea is just that some maximum level count is an easier invariant to maintain than ensuring that the value is some power of two less one. Not a big difference in the end though.